### PR TITLE
Synchronize Sage node title/CODAP attribute name changes

### PR DESCRIPTION
**Pull request by kswenson**
- synchronize nodes/attribute IDs and titles/names at initialization time
- serialize synchronized IDs and names/titles for continued synchronization
- call plugin API updateAttributes when node titles are changed [#134408929]
- respond to plugin API updateAttributes notification by changing node title when appropriate [#137900155]

Requires CODAP PR [#151](https://github.com/concord-consortium/codap/pull/151) to provide the necessary changes to the CODAP plugin API. Requires CODAP PR [#152](https://github.com/concord-consortium/codap/pull/152) to support undo/redo of title changes when synchronizing CODAP attribute names.

### CODE REVIEW

File: src/code/models/codap-connect.coffee [hunk 1]
```diff
@@ -77,10 +79,18 @@ module.exports = class CodapConnect
       # ret==null is indication of timeout, not an indication that the data set
       # doesn't exist.
       if !ret or ret.success
+        @_initialSyncAttributeProperties ret.values?.collections?[1].attrs
```
Comment (dougmartin) [inline src/code/models/codap-connect.coffee]: Are you missing an existential operator after the [1] like this:

`@_initialSyncAttributeProperties ret.values?.collections?[1]?.attrs`

File: src/code/models/codap-connect.coffee [hunk 2]
```diff
@@ -172,6 +184,52 @@ module.exports = class CodapConnect
       @codapPhone.call(getListing, doResolve)
       log.info "requested list of attributes"
 
+  _syncAttributeProperties: (attrProps, initialSync) ->
+    nodesToSync = if initialSync \
+                    then _.filter @graphStore.nodeKeys, (node) -> not node.codapID or not node.codapName \
+                    else _.map @graphStore.nodeKeys, (node) -> node
```
Comment (dougmartin) [inline src/code/models/codap-connect.coffee]: This `(node) -> node` looked weird at first but then I realized the _.map callback is `(value, key)` so it works.  It really does look like a noop though.

File: src/code/models/codap-connect.coffee (hunk 1 above)
Comment (kswenson) [inline src/code/models/codap-connect.coffee]: Good catch; I'll add it.

File: src/code/models/codap-connect.coffee (hunk 2 above)
Comment (kswenson) [inline src/code/models/codap-connect.coffee]: I added a comment that should make the intent somewhat clearer.

**Comment by kswenson** [comment]
@dougmartin One more commit (fixing undo/redo) to review.

**Comment by dougmartin** [comment]
Looks good 👍

**Pull request merged**